app/controllers/devise/two_factor_authentication_controller.rb

```diff
@@ -12,19 +13,32 @@ class TwoFactorAuthenticationController < DeviseController
     def new
       analytics.track_event('User requested a new OTP code')
 
-      current_user.send_new_otp
+      send_user_otp
+
+      show_direct_otp_prompt
       flash[:notice] = t('devise.two_factor_authentication.user.new_otp_sent')
-      redirect_to user_two_factor_authentication_path(method: 'sms')
     end
 
     def show
       if use_totp?
         show_totp_prompt
       else
-        show_direct_otp_prompt
+        @phone_number = user_decorator.masked_two_factor_phone_number
       end
     end
 
+    def confirm
+      @phone_number = user_decorator.masked_two_factor_phone_number
+      @code_value = current_user.direct_otp if FeatureManagement.prefill_otp_codes?
+    end
+
+    def send_code
+      analytics.track_event("User requested #{current_otp_delivery_method} OTP delivery")
+      send_user_otp
+      show_direct_otp_prompt
+      flash[:success] = t("notices.send_code.#{current_otp_delivery_method}")
+    end
```

Comment:
How the user invokes the sending of the OTP is different, sure, and there is already UI in place for that difference, but what happens when a user clicks "Confirm with text message" and "Resend" is exactly the same thing. We have 2 methods that do exactly the same thing. That's unnecessary duplication.

The 2 user actions you are talking about here are the 2 screens where they can trigger an OTP delivery (whether an initial one or by resending one that already happened). Those 2 user flows are already in place. Just because there are 2 ways to do the same thing, it doesn't mean we need 2 different methods that do the same thing, right?